### PR TITLE
feat: rescript parser and queries

### PR DESCRIPTION
**Pull request by ribru17**
Adds [rescript](https://rescript-lang.org/) parser. Queries adapted/improved from the base repo (which currently doesn't track `parser.c`, but there is a PR for it).

### CODE REVIEW

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1756]: Sorry, but that is not acceptable (anymore). If it doesn't build out of the box, it needs its own plugin.

**Review by clason — CHANGES_REQUESTED** [review]

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [reply lua/nvim-treesitter/parsers.lua:1756]: I don't think that's required, though -- but they do need to fix the include header for their scanner (and probably could use a bunch of other updates @amaanq)

Comment (ribru17) [reply lua/nvim-treesitter/parsers.lua:1756]: There is a PR to add all the generated files (and maybe it also fixes what you are talking about?). I'll mark as draft  and wait for that to get merged, that way no `generate_from_grammar` needed either

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:1756]: The `generate_from_grammar` is fine; it's the `use_makefile` that is unacceptable.

(And the PR does not fix the build issue since it doesn't touch the scanner.)

I would recommend making a parallel PR _just_ fixing that (and maybe committing the two json files and headers); I have total sympathy for maintainers noping out on a +200K diffstat PR. (The bindings were a mistake. `ts g --no-bindings` 4tw.)

Comment (ribru17) [reply lua/nvim-treesitter/parsers.lua]: Sounds good. Also is it possible the build issue is mac only? For me the parser builds when checking out this PR, and `ts g --no-bindings && ts t` works fine

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: Could be platform (compiler) specific, yes. But it still should be fixed sooner this would pick up a system treesitter header (which could be any old version).

**Comment by Emilios1995** [comment]
@ribru17 The queries and highlights on the main repo's `main` branch definitely need some improvement, but it would be great for those improvements to be on the parser's repo itself. We've recently moved the repo to `rescript`'s GH organization and are looking into trying to actively maintain it, so we will look at incoming PRs. 

Actually, we have been working on improving the queries on this [PR](https://github.com/rescript-lang/tree-sitter-rescript/pull/241/), which we're testing specifically on Neovim, so it would be even better if you made improvements over _that_.  (I haven't followed up on the PR for a few weeks, but I intend to do so and get it merged soon.)

**Comment by clason** [comment]
@ribru17 What's the status here? We "own" our queries (which are editor-specific) so we don't want to get blocked by slow upstream maintenance unless absolutely necessary.

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1755]:
```suggestion
```
That doesn't seem to be true? The repo contains an (up to date) `parser.c`.

**Comment by ribru17** [comment]
> @ribru17 What's the status here? We "own" our queries (which are editor-specific) so we don't want to get blocked by slow upstream maintenance unless absolutely necessary.

Just saw this, sorry. Upstream PR has been merged now thankfully, fixing this one up right now

File: queries/rescript/highlights.scm
Comment (clason) [inline queries/rescript/highlights.scm:7]: Can this be replaced with a Lua pattern?

**Comment by clason** [comment]
I worry a bit that these queries are overly specific at the cost of performance -- is this really necessary for acceptable highlighting?

File: queries/rescript/highlights.scm
Comment (lucario387) [inline queries/rescript/highlights.scm:141]: How does the tree look like under `pipe_expression`? This looks like it can become hellish very quickly

File: queries/rescript/highlights.scm
Comment (clason) [inline queries/rescript/highlights.scm:105]: 👀 
```suggestion
; first-level destructuring (required for Neovim as it only matches direct
; children and the above patterns do not match destructuring patterns)
```

👀  Can you comment on that?

**Comment by clason** [comment]
Testing `ts g` with HEAD CLI gives the dreaded
```
Warning: rule number is just a `seq` or `choice` rule with a single element. This is unnecessary.
```

This often (but not always) means that the generated parser does not correctly expose the corresponding node to the queries. That should be fixed before inclusion.

**Comment by ribru17** [comment]
@Emilios1995 would you mind commenting on some of these requests? You are more familiar with these queries

File: queries/rescript/highlights.scm
Comment (clason) [reply queries/rescript/highlights.scm:7]: (I think it can, so it should.)

Comment (ribru17) [reply queries/rescript/highlights.scm:7]: Yes I agree, if I resolved this without fixing it was by accident haha. Apologies

**Comment by ribru17** [comment]
> Testing `ts g` with HEAD CLI gives the dreaded
> 
> ```
> Warning: rule number is just a `seq` or `choice` rule with a single element. This is unnecessary.
> ```
> 
> This often (but not always) means that the generated parser does not correctly expose the corresponding node to the queries. That should be fixed before inclusion.

I remember when this issue was first floating around the root cause was kind of murky. Have there been false positives? Based on the grammar and the json file produced it seems like there are no states where `number` has a redundant `seq` or `choice`. In terms of the node being exposed in the grammar, it seems like it is - it exists in a lot of spots in the corpus tests at least

**Comment by clason** [comment]
>  Based on the grammar and the json file produced it seems like there are no states where number has a redundant seq or choice

There are no false positive reported so far, but it only reports the top rule even if the `seq` is in some other rule in a call sequence. In this case, the issue is https://github.com/rescript-lang/tree-sitter-rescript/blob/444c127686714b2358622427c3bdba7eb09021c6/grammar.js#L1367

**Comment by ribru17** [comment]
Ahh thanks I see, I'll make a PR

File: queries/rescript/highlights.scm
Comment (ribru17) [reply queries/rescript/highlights.scm:141]: Yeah, the queries were weird but I think all that was needed was an anchor here. I have split things up in a fixup commit to make them simpler, I think they are better now but let me know what you think

File: queries/rescript/highlights.scm
Comment (ribru17) [reply queries/rescript/highlights.scm:105]: Good catch, think I will just remove that comment. Looking through the queries+grammar looks like this comment is basically just trying to say that those queries are necessary because, by default, something like `(parent (child) @cap)` uses immediate parent matching and not something like `has-ancestor` to match `child`. But that goes without saying and this comment just serves to increase confusion I think

**Review by clason — APPROVED** [review]

**Comment by ribru17** [comment]
> I worry a bit that these queries are overly specific at the cost of performance -- is this really necessary for acceptable highlighting?

I have broken up and simplified some of the weirder queries, most are the same as you can see in the fixup commit but hopefully performance won't be too bad. Unless just the quantity of queries themselves has a non-negligible effect

**Comment by clason** [comment]
> Unless just the quantity of queries themselves has a non-negligible effect

It very much does, but I don't think these are excessive (but I haven't tested these myself). `check-queries` output will give you some idea.

File: queries/rescript/indents.scm
Comment (lucario387) [inline queries/rescript/indents.scm:28]: Is there some problem with merging these 2 queries?
```suggestion
(jsx_self_closing_element
  "/" @indent.branch
  ">"? @indent.end)
```

**Review by lucario387 — APPROVED** [review]

**Pull request merged**
